enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1120 @@
+
+```go
+// ServiceAliasConfig is a route for a service.  Uniquely identified by host + path.
+type ServiceAliasConfig struct {
+    ...
+    // For route based setting of http response headers
+    HTTPResponseHeaders []HTTPHeader
+
+    // For route based setting of http request headers
+    HTTPRequestHeaders []HTTPHeader
+}
+```
+```go
+// HTTPHeader specifies an HTTP header that should be set/replace/append
+type HTTPHeader struct {
+    // Name specifies an HTTP header name.
+    Name string
+
+    // Value specifies the header value.
+    Value string
+
+    // Action specifies the action to be performed.
+    Action string
+}
+```
+
+The sequence of action operations will be maintained according to the router RouteSpec. 
+
+First, we iterate  over `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type`,
+then over `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type`.
+
+C) library-go code will be having validations for checking Route API fields for OCP.
+   We will be creating an immutable/hard-coded deny-list of HTTP response headers in [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141)
+   which won't be allowed to be set or deleted. 
+
+   For the initial enhancement, some headers are in the deny list. 
+   Please refer to the section `Risks and Mitigations` for more details.
+   
+   Only 20 items in request and response header list will be allowed.
+   
+   Minimum length of the header value would have to be greater than 0 and less than 16384.
+   
+   Allowed sample fetchers for response dynamic header values are "res.hdr" and "ssl_c_der".
+
+   Allowed sample fetchers for request dynamic header values  are "req.hdr" and "ssl_c_der".
+   
+   Converters allowed for request and response are "lower" and "base64".
+
+   Validation prohibits configuring any header actions on passthrough routes.
+
+D) route/v1 of openshift/api will specify kubebuilder tags and CEL to enable client-side validation.
+
+E) operator/v1/types_ingress.go of openshift/api will contain the validations using kubebuilder tags and CEL.
+The validations will be the same as mentioned in point C) except the one for passthrough routes.
+
+F) To validate the header value it has to match a regex.
+For the IngressController CRD and Route CRD, this validation is defined on the request and response action lists in openshift/api.
+For the Route API as handled by openshift-apiserver, this is defined in library-go.
+This will be the regex:
+```go
+For request header:
+`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
+```
+```go
+For response header:
+`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
+```
```

Comment:
````suggestion
This will be the regex
for a request header value:
```
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
And for a response header value:
```
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
````

Reply:
done

Reply:
enhancements/ingress/set-delete-http-headers.md:665 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
enhancements/ingress/set-delete-http-headers.md:669 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
enhancements/ingress/set-delete-http-headers.md:899 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
enhancements/ingress/set-delete-http-headers.md:904 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]

Reply:
I need to specify a language.

Reply:
https://github.com/markdownlint/markdownlint/blob/main/docs/RULES.md#md040---fenced-code-blocks-should-have-a-language-specified suggests using ```` ```text ```` if there is no suitable language.  Alternatively, you can put a Go statement:
````suggestion
This will be the regex
for a request header value:
```go
permittedRequestHeaderValueRE = `^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
And for a response header value:
```go
permittedResponseHeaderValueRE = `^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
````